server/engine/rooms.js

```diff
@@ -1,41 +1,123 @@
-const users = require("./users");
+const Users = require("./users");
 
 /** Bijective hash on 32-bit ints */
-function hash(x) {
-  let h = x;
+function hashInt32(x) {
+  let h = x | 0;
   h = ((h >> 16) ^ h) * 0x45d9f3b;
   h = ((h >> 16) ^ h) * 0x45d9f3b;
-  h = (h >> 16) ^ h;
+  h ^= h >> 16;
   return h;
 }
 
 const nameLen = 4;
 
-/* Utility functions for the Rooms class. */
-function initUser(userId) {
-  return { userId, numConnections: 1 };
+class RoomUser {
+  constructor(id) {
+    this.id = id;
+    this.present = true;
+    this.spectate = false;
+    this.count = 0;
+  }
 }
 
-function getUserIds(userss) {
-  return userss.map(({ userId }) => userId);
-}
+class Room {
+  constructor(name, capacity = -1) {
+    this.name = name;
+    this.capacity = capacity;
+    this.users = [];
+    this.numPlayers = 0;
+    this.locked = false;
+  }
 
-function removeUser(userss, userId) {
-  return userss.filter(({ userId: id }) => id !== userId);
-}
+  canJoinAsPlayer() {
+    if (this.locked) return false;
+    return this.capacity < 0 || this.numPlayers < this.capacity;
+  }
+
+  join(userId) {
+    const index = this.users.findIndex(({ id }) => id === userId);
+    const user = index >= 0 ? this.users[index] : new RoomUser(userId);
+
+    if (index >= 0 && !user.present) {
+      // Splice user (and later repush) if it was inactive
+      this.users.splice(index, 1);
+    }
+    if (index < 0 || !user.present) {
+      // Push (or repush) user if it is new or was inactive
+      if (this.canJoinAsPlayer()) {
+        if (!user.spectate) this.numPlayers += 1;
+      } else {
+        user.spectate = true;
+      }
+      this.users.push(user);
+    }
+    user.present = true;
+    user.count += 1;
 
-function matchUserId(id) {
-  return ({ userId }) => userId === id;
+    return user;
+  }
+
+  leave(userId) {
+    const index = this.users.findIndex(({ id }) => id === userId);
+    if (index < 0) return undefined;
+    const user = this.users[index];
+    if (!user.present) return undefined;
```

Comment:
it might be a good idea to expand our definition for socket messages to be {type, payload, error} so that we can explicitly provide error messages as the error value rather than just returning null and making assumptions about what that means